openlibrary/core/bookshelves.py

```diff
@@ -202,6 +203,104 @@ def count_total_books_logged_by_user_per_shelf(
+    def add_storage_items_for_redirects(
+        cls, reading_log_keys, solr_docs: list[web.Storage]
+    ) -> list[web.storage]:
+        """
+        Use reading_log_keys to fill in missing redirected items in the
+        the solr_docs query results.
+
+        Solr won't return matches for work keys that have been redirected. Because
+        we use Solr to build the lists of storage items that ultimately gets passed
+        to the templates, redirected items returned from the reading log DB will
+        'disappear' when not returned by Solr. This remedies that by filling in
+        dummy works, albeit with the correct work_id.
+        """
+
+        from openlibrary.plugins.worksearch.code import run_solr_query
+        from openlibrary.plugins.worksearch.schemes.works import WorkSearchScheme
+
+        fetched_keys = {doc["key"] for doc in solr_docs}
+        missing_keys = {work for (work, _) in reading_log_keys} - fetched_keys
+
+        """
+        Provides a proper 1-to-1 connection between work keys and edition keys; needed, in order to fill in the appropriate 'logged_edition' data
+        for the correctly pulled work later on, as well as to correctly update the post-redirect version back to the pre_redirect key.
+        Without this step, processes may error due to the 'post-redirect key' not actually existing within a user's reading log.
+        """
+        work_to_edition_keys = {
+            work: edition for (work, edition) in reading_log_keys if edition
+        }
+        edition_to_work_keys = {
+            edition: work for (work, edition) in reading_log_keys if edition
+        }
+
+        # Here, we add in dummied works for situations in which there is no edition key present, yet the work key accesses a redirect.
+        # Ideally, this will be rare, as there's no way to access the relevant information through Solr.
+        for key in missing_keys.copy():
+            if not work_to_edition_keys.get(key):
+                missing_keys.remove(key)
+                solr_docs.append(web.storage({"key": key}))
+
+        edition_keys_to_query = [
+            work_to_edition_keys[key].split("/")[2] for key in missing_keys
+        ]
+        fq = f'edition_key:{" OR ".join(edition_keys_to_query)}'
+        if edition_keys_to_query:
+            solr_resp = run_solr_query(
+                scheme=WorkSearchScheme(),
+                param={'q': '*:*'},
+                rows=len(edition_keys_to_query),
+                fields=list(
+                    WorkSearchScheme.default_fetched_fields
+                    | {'subject', 'person', 'place', 'time', 'edition_key'}
+                ),
+                facet=False,
+                extra_params=[("fq", fq)],
+            )
+
+        """
+        Now, we add the correct 'logged_edition' information to each document retrieved by the query, and substitute the work_key in
+        each doc for the original one.
+        """
+        for doc in solr_resp.docs:
```

Comment:
This fails when `edition_keys_to_query` is an empty list.  Example [here](https://testing.openlibrary.org/people/l_via_d_upinov_/books?debug=true).

Reply:
Thanks! I'm not sure how I let something this simple slip through. It was more or less an indent error originally, but it should be resolved now.